File: SWEA/D3/2806. N－Queen/N－Queen.py
def solution():
    N = int(input())
    # i번째 행에 j열에 퀸을 둠, -1은 아직 미정 ex) [0, 1] -> 0행의 0열, 1행의 1열에 퀸을 둠
    row = [-1 for _ in range(N)]

    def validLocation(currentRow, col):
        for r in range(currentRow):
            if row[r] == col:  # 이전 행들 중 현재와 같은 열이 있다면 false
                return False

            if abs(row[r] - col) == abs(currentRow - r):  # 이전 행의 어떤 열과 현재 열의 차 == 이전행과 현재 행 번호와 같아면 대각선에 있는 것
                return False

        return True

    def dfs(currentRow):
        global result

        if currentRow == N:
            result += 1
            return
        for col in range(N):
            row[currentRow] = col
            if validLocation(currentRow, col):
                dfs(currentRow + 1)

    dfs(0)

    return result


if __name__ == '__main__':
    testCaseCount = int(input())
    for testCase in range(testCaseCount):
        result = 0
        print(f"#{testCase + 1} {solution()}")

File: 프로그래머스/2/12952. N－Queen/N－Queen.js
function solution(n) {
    // 1차원 배열로 생각
    // ex) [1, 3, 0, 2] 라면 1행은 1번, 2행은 3번... 인덱스에 존재
    const board = new Array(n).fill(-1)
    let result = 0
    
    function isValid(x, y) {
        for (let i=0; i<x; i++) {
            // 이미 존재하는 열이 있으면 불가
            if (board[i] === y) return false
            // 대각선 체크 -> index 차이가 n이라면, 두 값의 차도 n이어야 함(절대값)
            // [2, 3]
            if (Math.abs(x - i) === Math.abs(y - board[i])) return false
        }
        return true
    }   
    
    function dfs(x) {
        // 끝까지 도달했다면 결과 추가
        if (x === n) {
            result++
            return
        }
        
        for (let i=0; i<n; i++) {
            if (isValid(x, i)) {
                board[x] = i
                dfs(x+1, board)
                // 백트래킹 - 퀸 제거
                board[x] = -1
            }
        }
    }
    
     dfs(0)
    
    return result;
}